test-suite/src/test/scala/org/scalajs/testsuite/javalib/TreeSetTest.scala

```diff
@@ -0,0 +1,365 @@
+
+import scala.collection.JavaConversions._
+import scala.collection.mutable
+
+import org.scalajs.jasminetest.JasmineTest
+
+import scala.scalajs.runtime.UndefinedBehaviorError
+
+import java.{util => ju}
+import ju.TreeSet
+import ju.Comparator
+
+
+object TreeSetTest extends TreeSetTest(new TreeSetFactory)
+object TreeSetWithNullTest extends TreeSetTest(new TreeSetWithNullFactory)
+
+class TreeSetTest[F <: TreeSetFactory](treeSetTestFactory: F)
+    extends AbstractSetTest[F](treeSetTestFactory)
+    with SortedSetTest
+    with NavigableSetTest {
+
+  override def testApi(): Unit = {
+    super.testApi()
+    testSortedSetApi(treeSetTestFactory)
+    testNavigableSetApi(treeSetTestFactory)
+
+    it("should store and remove ordered integers") {
+      val ts = treeSetTestFactory.empty[Int]
+
+      expect(ts.size()).toEqual(0)
+      expect(ts.add(222)).toBeTruthy
+      expect(ts.size()).toEqual(1)
+      expect(ts.add(111)).toBeTruthy
+      expect(ts.size()).toEqual(2)
+      expect(ts.first).toEqual(111)
+      expect(ts.remove(111)).toBeTruthy
+
+      expect(ts.size()).toEqual(1)
+      expect(ts.first).toEqual(222)
+
+      expect(ts.remove(222)).toBeTruthy
+      expect(ts.size()).toEqual(0)
+      expect(ts.isEmpty).toBeTruthy
+      expect(ts.remove(333)).toBeFalsy
+      expect {
+        try {
+          ts.first
+          false
+        } catch {
+          case _: NoSuchElementException => true
+          case _: Throwable => false
+        }
+      }.toBeTruthy
+
+      if (treeSetTestFactory.allowsNullElement) {
+        expect(ts.add(null.asInstanceOf[Int])).toBeTruthy
+        expect(ts.contains(null)).toBeTruthy
+        expect(ts.remove(null)).toBeTruthy
+        expect(ts.contains(null)).toBeFalsy
+      }
+    }
+
+    it("should store and remove ordered strings") {
+      val ts = treeSetTestFactory.empty[String]
+
+      expect(ts.size()).toEqual(0)
+      expect(ts.add("222")).toBeTruthy
+      expect(ts.size()).toEqual(1)
+      expect(ts.add("111")).toBeTruthy
+      expect(ts.size()).toEqual(2)
+      expect(ts.first).toEqual("111")
+      expect(ts.remove("111")).toBeTruthy
+
+      expect(ts.size()).toEqual(1)
+      expect(ts.first).toEqual("222")
+
+      expect(ts.remove("222")).toBeTruthy
+      expect(ts.size()).toEqual(0)
+      expect(ts.remove("333")).toBeFalsy
+      expect(ts.isEmpty).toBeTruthy
+
+      if (treeSetTestFactory.allowsNullElement) {
+        expect(ts.add(null)).toBeTruthy
+        expect(ts.contains(null)).toBeTruthy
+        expect(ts.remove(null)).toBeTruthy
+        expect(ts.contains(null)).toBeFalsy
+      }
+    }
+
+    case class TestObj(num: Int)
+
+    when("compliant-asinstanceofs").
+    it("should throw exception on non comparable objects") {
+      val ts = treeSetTestFactory.empty[TestObj]
+
+      expect(ts.size()).toEqual(0)
+      expectThrows[ClassCastException](ts.add(TestObj(111)))
+      expect(ts.comparator eq null).toBeTruthy
```

Comment:
good catch the check will be done on the comparator null check result
```
if (ts.comparator eq null) {
...
} else {
...
}
```
if the condition is satisfied the inner tests must pass without the needs of the next boolean flag